### PR TITLE
efi: Add internal helpers for shim and inspecting PE binaries

### PR DESCRIPTION
**Pull request by chrisccoulson**
This adds some internal helpers for inspecting PE binaries and some shim
helpers as well - one of these replaces an existing one
(shimImageHandle).

These are all going to be used in a future PR, although the internal
shimImageHandle and peImageHandle are used in the existing code.

### CODE REVIEW

File: tools/make-efi-testdata/apps.go
Comment (pedronis) [inline tools/make-efi-testdata/apps.go:1]: these changes result in the mock* binary file changes and the test hash changes?

File: efi/pe.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+type sbatComponent struct {
+	Name              string
+	Generation        int
+	VendorName        string
+	VendorPackageName string
+	VendorVersion     string
+	VendorUrl         string
+}
+
+// peImageHandle provides utilities for working with a PE image.
```
Comment (pedronis) [inline efi/pe.go:46]: with a PE image's sections and signatures.

File: efi/pe.go [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	r.FieldsPerRecord = 6
+
+	var components []sbatComponent
+	for {
+		record, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, xerrors.Errorf("invalid record: %w", err)
```
Comment (pedronis) [inline efi/pe.go:139]: sbat record?

File: efi/pe.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+		}
+
+		gen, err := strconv.Atoi(record[1])
+		if err != nil {
+			return nil, xerrors.Errorf("invalid component generation: %w", err)
```
Comment (pedronis) [inline efi/pe.go:144]: sbat component?

File: efi/pe.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+		// Signatures in this section are 8-byte aligned - see the PE spec:
+		// https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#the-attribute-certificate-table-image-only
+		off, _ := certReader.Seek(0, io.SeekCurrent)
+		alignSize := (8 - (off & 7)) % 8
+		certReader.Seek(alignSize, io.SeekCurrent)
+
+		c, err := efi.ReadWinCertificate(certReader)
+		switch {
+		case xerrors.Is(err, io.EOF):
+			noMoreCerts = true
```
Comment (pedronis) [inline efi/pe.go:214]: this could use a "break Label" statement instead of a var

File: efi/shim.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+	shimVendorCertIsDb
+)
+
+// shimSbatLevel corresponds to the SBAT revocation level payload stored inside
+// shim.
+type shimSbatLevel [2][]byte
+
+// ForPolicy returns the SBAT revocation level for the specified policy.
+func (l shimSbatLevel) ForPolicy(policy shimSbatPolicy) []byte {
+	return l[policy-1]
```
Comment (pedronis) [inline efi/shim.go:225]: should this produce a more meaningful panic if policy is not 1 (latest) or 2  (previous) ?

File: efi/shim.go [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+			},
+		}, shimVendorCertIsX509, nil
+	}
+
+	// The vendor DB data contains one or more ESLs
+	db, err := efi.ReadSignatureDatabase(bytes.NewReader(dbData))
+	if err != nil {
+		return nil, 0, err
+	}
+	return db, shimVendorCertIsDb, nil
```
Comment (pedronis) [inline efi/shim.go:314]: this path is not exercised

**Review by pedronis — COMMENTED** [review]
did a pass, some comments

File: tools/make-efi-testdata/apps.go
Comment (chrisccoulson) [reply tools/make-efi-testdata/apps.go:1]: They do

File: efi/pe.go (hunk 1 above)
Comment (chrisccoulson) [reply efi/pe.go]: I've updated this now.

File: efi/pe.go (hunk 2 above)
Comment (chrisccoulson) [reply efi/pe.go]: I've updated this.

File: efi/pe.go (hunk 3 above)
Comment (chrisccoulson) [reply efi/pe.go]: I've updated this.

File: efi/pe.go (hunk 4 above)
Comment (chrisccoulson) [reply efi/pe.go]: I've changed this to use a label.

File: efi/shim.go (hunk 5 above)
Comment (chrisccoulson) [reply efi/shim.go:228]: I've added a check here so that it panics with a useful message.

File: efi/shim.go (hunk 6 above)
Comment (chrisccoulson) [reply efi/shim.go:317]: I've added a test case for this now.

**Review by pedronis — APPROVED** [review]
thanks

**Pull request merged**